File: scripts/bindings/ggplot_discrete_secondary.py
"""FIX-GG04: discrete identity duplicate axes through the actual Python adapter."""
import json, math, sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes())
options=c.export_options(640,360).dpi(96).basis('current')
cases=json.loads((ROOT/'fixtures/parity/ggplot2/discrete-secondary.json').read_text())['cases'];records=[]
def key(v):return 'Null' if v is None else {'Text':v}
def semantic(v):return {'MissingCategory':None} if v is None else {'Category':v}
for index,case in enumerate(cases):
 for family in ('band','point'):
  owned=[]
  try:
   values=case['inputs'];control=case['control'];expected=case['result']
   data=c.Data.columns({'x':c.categorical([v or '' for v in values]).validity([v is not None for v in values]),'y':c.column([1.]*len(values),kind='float64')});owned.append(data)
   primary=c.x_axis().scale(c.scale_band() if family=='band' else c.scale_point()).range(100,540).discrete_policy({'limits':None if case['limits'] is None else list(map(key,case['limits'])),'na_translate':case['na_translate']}).guide_geometry({'labels':'Preserve'})
   if control in ('primary_breaks','primary_labels','primary_hidden'):
    guide={'breaks':[{'Text':'b'},{'Text':'a'}]} if control!='primary_hidden' else {'labels':'Hidden'}
    if control=='primary_labels':guide['labels']={'Explicit':['Bee','Aye']}
    primary=primary.discrete_policy({'limits':None if case['limits'] is None else list(map(key,case['limits'])),'na_translate':case['na_translate'],'guide':guide})
   secondary=c.x_axis().name('secondary').side('Top').secondary('x',2. if control=='transformed' else 1.,0.).guide_geometry({'labels':'Preserve'})
   if control=='empty':secondary=secondary.tick_values([])
   if control=='numeric':secondary=secondary.tick_values([{'Number':v} for v in [-1,0,.5,1,1.5,2,3,4,5]])
   if control in ('character','explicit','bad_labels'):
    keys=['b',None,'NA','absent','a'] if control=='character' else ['b',None,'NA','a'] if control=='explicit' else ['b','a']
    secondary=secondary.tick_values(list(map(semantic,keys)))
   if control in ('explicit','bad_labels','hidden'):
    labels=['Bee','Missing','Literal','A'] if control=='explicit' else ['wrong'] if control=='bad_labels' else ['']*len(expected.get('breaks',[]))
    secondary=secondary.tick_format({'Labels':labels})
   plot=c.plot(data).profile('Ggplot2_4_0_3').aes(c.aes().x('x').y('y')).layer(c.points()).x_axis(primary).x_axis(secondary).build();owned.append(plot)
   restored=c.Plot.from_json(plot.to_json());owned.append(restored)
   request=output.request(restored,options);owned.append(request);frame=request.prepare();owned.append(frame)
   assert 'error' not in expected,(index,expected)
   ticks=next(g for g in frame.guides()['guides'] if g['spec']['side']=='Top')['ticks']
   assert len(ticks)==len(expected['breaks']),(index,ticks,expected)
   for tick,value,label,position in zip(ticks,expected['breaks'],(['']*len(ticks) if control in ('hidden','primary_hidden') else expected['labels']),expected['positions']):
    assert math.isclose(tick['value']['Number'],value,rel_tol=0,abs_tol=1e-12)
    assert tick['label']==('NA' if label is None else label),(index,tick,label)
    assert math.isclose((tick['position']-100)/440,position,rel_tol=0,abs_tol=1e-12),(index,tick,position)
   records.append({'index':index,'family':family,'ticks':ticks})
   if index in (11,14,16) and family=='band':
    for fmt in ('svg','pdf','png'):(out/f'discrete-secondary-{control}.{fmt}').write_bytes(frame.export(fmt))
  except c.ChartError as error:
   diagnostic=json.loads(str(error));assert 'error' in case['result'],(index,diagnostic)
   records.append({'index':index,'family':family,'error':diagnostic['code']})
  finally:
   for value in reversed(owned):value.dispose()
assert len(records)==352
(out/'records.json').write_text(json.dumps(records,indent=2));options.dispose();output.dispose()
print('PASS Python: 352 discrete secondary configurations and nine publication files.')

File: crates/chart-core/tests/ggplot_discrete_secondary.rs
//! FIX-GG04 / GG2-03: primary discrete duplicate axes against the pinned R oracle.
use chart_core::{
    ChartResult, Rect, ResourceId, Revision, composition::ScaleValue, layout::*, prelude::*,
    scales::*, services::*,
};
use serde_json::Value;
struct Metrics;
impl TextMeasurer for Metrics {
    fn measure(&self, r: TextRequest<'_>) -> ChartResult<TextMetrics> {
        TextMetrics::new(r.text.len() as f64 * 5., 8., 2.)
    }
}
#[test]
fn discrete_secondary_matches_reference_panels() {
    let fixture: Value = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/discrete-secondary.json"
    ))
    .unwrap();
    let cases = fixture["cases"].as_array().unwrap();
    assert_eq!(cases.len(), 176);
    let request = LayoutRequest::new(
        Rect::new(0., 0., 640., 360.).unwrap(),
        Units::LogicalPixels,
        ResourceDescriptor {
            id: ResourceId::new(1),
            revision: Revision::INITIAL,
            kind: ResourceKind::Font,
            byte_len: 1,
        },
    );
    for case in cases {
        for point in [false, true] {
            let inputs = case["inputs"].as_array().unwrap();
            let data = Data::columns()
                .column(
                    "x",
                    categorical(inputs.iter().map(|v| v.as_str().unwrap_or("")))
                        .validity(inputs.iter().map(|v| !v.is_null()).collect()),
                )
                .column("y", vec![1.; inputs.len()])
                .build()
                .unwrap();
            let control = case["control"].as_str().unwrap();
            let mut secondary = x_axis()
                .name("secondary")
                .side(AxisSide::Top)
                .guide_geometry(Some(GuideGeometry {
                    labels: Some(GuideLabelPolicy::Preserve),
                    ..Default::default()
                }))
                .secondary("x", if control == "transformed" { 2. } else { 1. }, 0.);
            let keys = |values: &[Option<&str>]| {
                values
                    .iter()
                    .map(|v| {
                        v.map_or(ScaleValue::MissingCategory, |s| {
                            ScaleValue::Category(s.into())
                        })
                    })
                    .collect()
            };
            secondary =
                match control {
                    "empty" => secondary.tick_values(Some(vec![])),
                    "numeric" => secondary.tick_values(Some(
                        [-1., 0., 0.5, 1., 1.5, 2., 3., 4., 5.]
                            .into_iter()
                            .map(ScaleValue::Number)
                            .collect(),
                    )),
                    "character" => secondary.tick_values(Some(keys(&[
                        Some("b"),
                        None,
                        Some("NA"),
                        Some("absent"),
                        Some("a"),
                    ]))),
                    "explicit" => secondary
                        .tick_values(Some(keys(&[Some("b"), None, Some("NA"), Some("a")])))
                        .tick_format(Some(GuideFormatter::Labels(
                            ["Bee", "Missing", "Literal", "A"].map(String::from).into(),
                        ))),
                    "bad_labels" => secondary
                        .tick_values(Some(keys(&[Some("b"), Some("a")])))
                        .tick_format(Some(GuideFormatter::Labels(vec!["wrong".into()]))),
                    "hidden" => secondary.tick_format(Some(GuideFormatter::Labels(vec![
                            String::new();
                            case["result"]["breaks"].as_array().map_or(0, Vec::len)
                        ]))),
                    _ => secondary,
                };
            let mut policy = GgplotDiscretePosition {
                limits: case["limits"].as_array().map(|v| {
                    v.iter()
                        .map(|k| {
                            k.as_str()
                                .map_or(ScaleKey::Null, |s| ScaleKey::Text(s.into()))
                        })
                        .collect()
                }),
                na_translate: case["na_translate"].as_bool().unwrap(),
                ..Default::default()
            };
            if matches!(control, "primary_breaks" | "primary_labels") {
                policy.guide.breaks =
                    Some(vec![ScaleKey::Text("b".into()), ScaleKey::Text("a".into())]);
            }
            if control == "primary_labels" {
                policy.guide.labels =
                    GgplotGuideLabels::Explicit(vec![Some("Bee".into()), Some("Aye".into())]);
            }
            if control == "primary_hidden" {
                policy.guide.labels = GgplotGuideLabels::Hidden;
            }
            let figure = plot(data)
                .profile(Profile::Ggplot2_4_0_3)
                .aes(aes().x("x").y("y"))
                .layer(points())
                .x_axis(
                    x_axis()
                        .scale(if point { scale_point() } else { scale_band() })
                        .range(0., 100.)
                        .discrete_policy(Some(policy)),
                )
                .x_axis(secondary)
                .build()
                .unwrap();
            let figure = Plot::from_json(&figure.to_json().unwrap()).unwrap();
            let id = figure.axis("secondary").unwrap().id();
            let result = layout(
                figure.chart().unwrap().prepare().unwrap(),
                &request,
                &Metrics,
            );
            if !case["result"]["error"].is_null() {
                assert!(result.is_err(), "{case}");
                continue;
            }
            let frame = result.unwrap_or_else(|e| panic!("{e:?} {case}"));
            let axis = &frame.axes()[&id];
            let expected = &case["result"];
            assert_eq!(
                axis.ticks.len(),
                expected["breaks"].as_array().unwrap().len(),
                "{case}"
            );
            for (i, tick) in axis.ticks.iter().enumerate() {
                let ScaleValue::Number(value) = tick.value else {
                    panic!("numeric secondary index")
                };
                assert!(
                    (value - expected["breaks"][i].as_f64().unwrap()).abs() < 2e-14,
                    "{case}"
                );
                let label = if matches!(control, "hidden" | "primary_hidden") {
                    ""
                } else {
                    expected["labels"][i].as_str().unwrap_or("NA")
                };
                assert_eq!(tick.label, label, "{case}");
                assert!(
                    (tick.position / 100. - expected["positions"][i].as_f64().unwrap()).abs()
                        < 2e-14,
                    "position {} {case}",
                    tick.position / 100.
                );
            }
        }
    }
}

#[test]
fn discrete_secondary_uses_primary_spacing_in_both_orientations_and_directions() {
    let request = LayoutRequest::new(
        Rect::new(0., 0., 640., 360.).unwrap(),
        Units::LogicalPixels,
        ResourceDescriptor {
            id: ResourceId::new(1),
            revision: Revision::INITIAL,
            kind: ResourceKind::Font,
            byte_len: 1,
        },
    );
    for horizontal in [true, false] {
        for reverse in [true, false] {
            for point in [true, false] {
                let data = Data::columns()
                    .column("category", categorical(["a", "b", "c"]))
                    .column("number", vec![1., 2., 3.])
                    .build()
                    .unwrap();
                let range = if reverse { [540., 100.] } else { [100., 540.] };
                let primary = if horizontal { x_axis() } else { y_axis() }
                    .scale(if point { scale_point() } else { scale_band() })
                    .range(range[0], range[1]);
                let secondary = if horizontal { x_axis() } else { y_axis() }
                    .name("secondary")
                    .side(if horizontal {
                        AxisSide::Top
                    } else {
                        AxisSide::Right
                    })
                    .secondary(if horizontal { "x" } else { "y" }, 1., 0.)
                    .tick_format(Some(GuideFormatter::Numeric(Box::new(
                        chart_core::typography::NumericFormat {
                            specifier: ".2f".into(),
                            locale: Default::default(),
                        },
                    ))))
                    .guide_geometry(Some(GuideGeometry {
                        labels: Some(GuideLabelPolicy::Preserve),
                        ..Default::default()
                    }));
                let builder = plot(data).profile(Profile::Ggplot2_4_0_3).layer(points());
                let figure = if horizontal {
                    builder
                        .aes(aes().x("category").y("number"))
                        .x_axis(primary)
                        .x_axis(secondary)
                } else {
                    builder
                        .aes(aes().x("number").y("category"))
                        .y_axis(primary)
                        .y_axis(secondary)
                }
                .build()
                .unwrap();
                let id = figure.axis("secondary").unwrap().id();
                let frame = layout(
                    figure.chart().unwrap().prepare().unwrap(),
                    &request,
                    &Metrics,
                )
                .unwrap();
                let axis = &frame.axes()[&id];
                assert_eq!(axis.ticks.len(), 3);
                for (i, expected) in [0.187, 0.5, 0.812].into_iter().enumerate() {
                    let tick = axis
                        .ticks
                        .iter()
                        .find(|t| t.value == ScaleValue::Number(1. + i as f64))
                        .unwrap();
                    assert!(
                        ((tick.position - range[0]) / (range[1] - range[0]) - expected).abs()
                            < 2e-14
                    );
                    assert_eq!(tick.label, format!("{:.2}", 1. + i as f64));
                    assert!(axis.invert_value(tick.position).is_err());
                }
            }
        }
    }
}
